In [1]:


import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np 





## Composition Consumption and Cost Calculation

To accurately calculate the composition consumption and its corresponding cost for a given grade of steel, we follow a two-step process. This approach first determines the actual amount of composition used in kilograms (kg) and then calculates the cost based on the price per ton.

### Step 1: Calculate Composition Consumption in Kilograms (kg)

Given variables:
- `C` = Composition consumption rate in kg of composition per ton of billets
- `T` = Tons of billets produced

The formula to calculate the actual consumption of the composition (`A`) in kilograms (kg) for that grade of steel is:

$$\[A = C \times T\]$$

### Step 2: Convert Consumption into Cost

Given additional variable:
$$ `P` = Price of the sheet (or composition) per ton$$

Since the price `P` is provided per ton, and there are 1000 kg in a ton, we first convert the actual consumption from kilograms to tons (`A_tons`):

$$\[A_{\text{tons}} = \frac{A}{1000}\]$$

To calculate the total cost (`Cost`) of the composition based on its consumption and the sheet price per ton, we use the formula:

$$\[Cost = A_{\text{tons}} \times P\]$$

Combining the steps, we get the comprehensive formula for calculating cost:

$$\[Cost = \left( \frac{C \times T}{1000} \right) \times P\]$$

This formula allows us to calculate the total cost of the composition used based on its consumption rate per ton of billets produced, the total tons of billets produced, and the price per ton of the composition.
 
composition consumption in kg/ton * tons of billets produced = composition consumption in KG 
 1.   kg/ton * ton = kg you get the actual consumption of composition for that grade steel 
 2.   from 1, divide by 1000 to get ton, multiply the result by tth sheet price ton * (sheet price since its per ton) per composition.



## Prices for SMP2 Chemicals:

In [2]:


# Given dictionaries
list_names_prices_smp2_december_2023 = {
    "FeSi - Ferro-Silicon": 1784.26869542229,
    "FeSiMn - Silicon-maganese": 982.491100414191,
    "FeMnHC - Ferro-Manganese (High Carbon)": 1072.38667899182,
    "Recarburizer - Recarburizer": 459.168109303317,
    'FeCr - Ferrochrome': 0.0
}

list_names_prices_smp2_februrary_2024 = {
    "FeSi - Ferro-Silicon": 1513.14266939651,
    "FeSiMn - Silicon-maganese": 981.166637222105,
    "FeMnHC - Ferro-Manganese (High Carbon)": 987.371168766638,
    "Recarburizer - Recarburizer": 396.65551241767,
    'FeCr - Ferrochrome': 0.0
}

list_names_prices_smp2_january_2024 = {
    "FeSi - Ferro-Silicon": 1761.04858264133,
    "FeSiMn - Silicon-maganese": 1761.04858264133,
    "FeMnHC - Ferro-Manganese (High Carbon)": 1013.54162166185,
    "Recarburizer - Recarburizer": 485.931625802038,
    'FeCr - Ferrochrome': 0.0
}

list_names_prices_smp2_november_2023 = {
    "FeSi - Ferro-Silicon": 1813.81793731061,
    "FeSiMn - Silicon-maganese": 1068.45538821339,
    "FeMnHC - Ferro-Manganese (High Carbon)": 1072.38820871572,
    "Recarburizer - Recarburizer": 541.481184711259,
    'FeCr - Ferrochrome': 0.0
}

df = pd.DataFrame([
    list_names_prices_smp2_december_2023,
    list_names_prices_smp2_februrary_2024,
    list_names_prices_smp2_january_2024,
    list_names_prices_smp2_november_2023
])

mean_prices = df.mean()
list_names_prices_smp2_mean = mean_prices.to_dict()
list_names_prices_smp2_mean


{'FeSi - Ferro-Silicon': 1718.069471192685,
 'FeSiMn - Silicon-maganese': 1198.290427122754,
 'FeMnHC - Ferro-Manganese (High Carbon)': 1036.421919534007,
 'Recarburizer - Recarburizer': 470.809108058571,
 'FeCr - Ferrochrome': 0.0}

In [3]:

# FeSiMn - Silicon-maganese average cost over 4 months
(982.491100414191 + 981.166637222105 + 1761.04858264133 +  1068.45538821339) /4



1198.290427122754

In [4]:

# FeSi - Ferro-Silicon average cost per 4 months:

(1784.26869542229 + 1513.14266939651 + 1761.04858264133 + 1813.81793731061) / 4


1718.069471192685

In [5]:

# choose SMP 
data = pd.read_csv("Ferroalloys Consumption SMP #2.csv")



# extract columns you are going to use for the analysis:
data = data[['SteelGrade', 
             'Billet',
             'FeSiMn Target (KG/Ton)',
             'FeSiMn Actual (KG/Ton)',
             'FeSi Target (KG/Ton)',
             'FeSi Actual (KG/Ton)', 
             ]]




# calculate overshot per Chemicals:
data['FeSiMn Overshot (KG/Ton)'] = data['FeSiMn Actual (KG/Ton)'] - data['FeSiMn Target (KG/Ton)']
data['FeSi Overshot (KG/Ton)'] = data['FeSi Actual (KG/Ton)'] - data['FeSi Target (KG/Ton)']


# separate dataframe to do the calculations, to be more accurate and precise:
fesmin_data = data[["SteelGrade","Billet","FeSiMn Actual (KG/Ton)","FeSiMn Target (KG/Ton)","FeSiMn Overshot (KG/Ton)"]]
fesi_data = data[["SteelGrade","Billet","FeSi Actual (KG/Ton)","FeSi Target (KG/Ton)","FeSi Overshot (KG/Ton)"]]




In [6]:

# now calculate here conservative estimates and compare them with the base:

data_conservative = pd.read_csv("Ferroalloys Consumption SMP #2.csv")



# extract columns you are going to use for the analysis:
data_conservative = data_conservative[['SteelGrade',
             'Billet',
             'FeSiMn Target (KG/Ton)',
             'FeSiMn Actual (KG/Ton)',
             'FeSi Target (KG/Ton)',
             'FeSi Actual (KG/Ton)',
             ]]


data_conservative["FeSiMn Target (KG/Ton)"] = data_conservative["FeSiMn Target (KG/Ton)"] * 1.025
data_conservative["FeSi Target (KG/Ton)"] = data_conservative["FeSi Target (KG/Ton)"] * 1.025
 


# now calculate overshot per mentioned chemicals:

data_conservative['FeSiMn Overshot (KG/Ton)'] = data_conservative['FeSiMn Actual (KG/Ton)'] - data_conservative['FeSiMn Target (KG/Ton)']
data_conservative['FeSi Overshot (KG/Ton)'] = data_conservative['FeSi Actual (KG/Ton)'] - data_conservative['FeSi Target (KG/Ton)']



fesmin_data_conservative = data_conservative[["SteelGrade","Billet","FeSiMn Actual (KG/Ton)","FeSiMn Target (KG/Ton)","FeSiMn Overshot (KG/Ton)"]]
fesi_data_conservative = data_conservative[["SteelGrade","Billet","FeSi Actual (KG/Ton)","FeSi Target (KG/Ton)","FeSi Overshot (KG/Ton)"]]



### Rename columns to standardize them:

In [7]:

fesmin_data = fesmin_data.rename(columns={'Billet': 'Billet Mass (Tons)'})
fesi_data =  fesi_data.rename(columns={'Billet': 'Billet Mass (Tons)'})


In [8]:


fesmin_data_conservative = fesmin_data_conservative.rename(columns={'Billet': 'Billet Mass (Tons)'})
fesi_data_conservative =  fesi_data_conservative.rename(columns={'Billet': 'Billet Mass (Tons)'})



### Add Suffix to duplicated steel grades names:

In [9]:

fesi_data['Suffix'] = fesi_data.groupby('SteelGrade').cumcount() + 1
fesi_data['SteelGrade_Unique'] = fesi_data['SteelGrade'] + '-' + fesi_data['Suffix'].astype(str)
fesi_data["SteelGrade"] = fesi_data["SteelGrade_Unique"]
fesi_data.drop(columns=['Suffix',"SteelGrade_Unique"], inplace=True)



In [10]:
# add unique name to steel grade

fesmin_data['Suffix'] = fesmin_data.groupby('SteelGrade').cumcount() + 1
fesmin_data['SteelGrade_Unique'] = fesmin_data['SteelGrade'] + '-' + fesmin_data['Suffix'].astype(str)
fesmin_data["SteelGrade"] = fesmin_data["SteelGrade_Unique"]
fesmin_data.drop(columns=['Suffix',"SteelGrade_Unique"], inplace=True)


In [11]:

fesi_data_conservative['Suffix'] = fesi_data_conservative.groupby('SteelGrade').cumcount() + 1
fesi_data_conservative['SteelGrade_Unique'] = fesi_data_conservative['SteelGrade'] + '-' + fesi_data_conservative['Suffix'].astype(str)
fesi_data_conservative["SteelGrade"] = fesi_data_conservative["SteelGrade_Unique"]
fesi_data_conservative.drop(columns=['Suffix',"SteelGrade_Unique"], inplace=True)



In [12]:

fesmin_data_conservative['Suffix'] = fesmin_data_conservative.groupby('SteelGrade').cumcount() + 1
fesmin_data_conservative['SteelGrade_Unique'] = fesmin_data_conservative['SteelGrade'] + '-' + fesmin_data_conservative['Suffix'].astype(str)
fesmin_data_conservative["SteelGrade"] = fesmin_data_conservative["SteelGrade_Unique"]
fesmin_data_conservative.drop(columns=['Suffix',"SteelGrade_Unique"], inplace=True)


### Filter columns where target must not equal 0 and overshot must be greater than 0.0:

In [13]:

# filter columns you are interested in:
# first condition target must be greater than 0.0:
fesmin_data = fesmin_data[fesmin_data["FeSiMn Target (KG/Ton)"]!= 0.0]

# second condition overshot must be > 0.0 
fesmin_data = fesmin_data[fesmin_data["FeSiMn Overshot (KG/Ton)"] > 0.0]



In [14]:



# filter columns you are interested in:
# first condition target must be greater than 0.0:
fesmin_data_conservative = fesmin_data_conservative[fesmin_data_conservative["FeSiMn Target (KG/Ton)"]!= 0.0]

# second condition overshot must be > 0.0 
fesmin_data_conservative = fesmin_data_conservative[fesmin_data_conservative["FeSiMn Overshot (KG/Ton)"] > 0.0]




In [15]:

# filter columns you are interested in:
# first condition target must be greater than 0.0:
fesi_data = fesi_data[fesi_data["FeSi Target (KG/Ton)"]!= 0.0]

# second condition overshot must be > 0.0 
fesi_data = fesi_data[fesi_data["FeSi Overshot (KG/Ton)"] > 0.0]



In [16]:


# filter columns you are interested in:
# first condition target must be greater than 0.0:
fesi_data_conservative = fesi_data_conservative[fesi_data_conservative["FeSi Target (KG/Ton)"]!= 0.0]

# second condition overshot must be > 0.0 
fesi_data_conservative = fesi_data_conservative[fesi_data_conservative["FeSi Overshot (KG/Ton)"] > 0.0]




In [17]:

fesmin_data


SteelGrade  Billet Mass (Tons)  FeSiMn Actual (KG/Ton)  \
0          3SP CON-1           10040.952                    7.79   
2          3SP UIS-1           19396.407                    8.14   
8    AISI 1006 CAQ-1           14217.135                    3.14   
10       AISI 1012-1             127.478                    7.89   
11       AISI 1012-2            2847.713                    7.73   
12       AISI 1015-1            3129.471                    8.02   
13       AISI 1018-1            8659.960                    8.91   
14       AISI 1038-1            2732.942                   10.07   
20    BS-G500B R22-1           11144.716                    7.92   
21    BS-G500B R22-2           11956.901                    8.00   
22    BS-G500B R51-1            1242.724                   10.37   
23      BSG500BR41-1           74008.489                    7.98   
24      BSG500BR41-2          179322.080                    7.91   
25      BSG500BR42-1            1980.089                    8.04   
26      BSG500BR42-2          151789.712                    7.83   
27      BSSS560R12-1            8578.967                   11.86   
28      BSSS560R22-1            2217.466                    7.72   
29      BSSS560R22-2            7713.394                    7.97   
32  KWSB500BWR R11-1            2948.755                    8.70   
33  KWSB500BWR R12-1            1364.831                    9.84   
34  KWSB500BWR R22-1           14175.175                    7.67   
35       S235 JR-A-1            4384.918                   10.57   
36       S235 JR-A-2           19308.201                   10.32   
37       S275 JR-A-1           23003.444                   14.64   
38       S275 JR-A-2           75502.051                   14.70   
39     S275 JR-AAA-1           12155.000                   14.42   
40     S275 JR-AAA-2          112660.245                   14.53   
41      S355 G11-A-1            1886.049                   18.52   
42      S355 G11-A-2            5001.665                   17.10   
57        SAE 1012-1             129.499                    8.66   
58        SAE 1018-1            3862.805                   10.04   
74         SWRY 11-1           14536.372                    2.45   

    FeSiMn Target (KG/Ton)  FeSiMn Overshot (KG/Ton)  
0                      6.0                      1.79  
2                      6.2                      1.94  
8                      1.5                      1.64  
10                     6.5                      1.39  
11                     6.5                      1.23  
12                     5.8                      2.22  
13                     8.2                      0.71  
14                    10.0                      0.07  
20                     5.9                      2.02  
21                     5.9                      2.10  
22                     7.4                      2.97  
23                     5.9                      2.08  
24                     5.9                      2.01  
25                     7.4                      0.64  
26                     7.4                      0.43  
27                     9.0                      2.86  
28                     5.9                      1.82  
29                     5.9                      2.07  
32                     8.5                      0.20  
33                     9.0                      0.84  
34                     5.9                      1.77  
35                     9.0                      1.57  
36                     9.0                      1.32  
37                    13.6                      1.04  
38                    13.6                      1.10  
39                    13.6                      0.82  
40                    13.6                      0.93  
41                    16.4                      2.12  
42                    16.4                      0.70  
57                     6.5                      2.16  
58                     8.2                      1.84  
74   

In [18]:

# we miss two steel grade when we change the target to increase by 2.5% 
# AISI 1038-1
# KWSB500BWR R11-1


fesmin_data_conservative


SteelGrade  Billet Mass (Tons)  FeSiMn Actual (KG/Ton)  \
0          3SP CON-1           10040.952                    7.79   
2          3SP UIS-1           19396.407                    8.14   
8    AISI 1006 CAQ-1           14217.135                    3.14   
10       AISI 1012-1             127.478                    7.89   
11       AISI 1012-2            2847.713                    7.73   
12       AISI 1015-1            3129.471                    8.02   
13       AISI 1018-1            8659.960                    8.91   
20    BS-G500B R22-1           11144.716                    7.92   
21    BS-G500B R22-2           11956.901                    8.00   
22    BS-G500B R51-1            1242.724                   10.37   
23      BSG500BR41-1           74008.489                    7.98   
24      BSG500BR41-2          179322.080                    7.91   
25      BSG500BR42-1            1980.089                    8.04   
26      BSG500BR42-2          151789.712                    7.83   
27      BSSS560R12-1            8578.967                   11.86   
28      BSSS560R22-1            2217.466                    7.72   
29      BSSS560R22-2            7713.394                    7.97   
33  KWSB500BWR R12-1            1364.831                    9.84   
34  KWSB500BWR R22-1           14175.175                    7.67   
35       S235 JR-A-1            4384.918                   10.57   
36       S235 JR-A-2           19308.201                   10.32   
37       S275 JR-A-1           23003.444                   14.64   
38       S275 JR-A-2           75502.051                   14.70   
39     S275 JR-AAA-1           12155.000                   14.42   
40     S275 JR-AAA-2          112660.245                   14.53   
41      S355 G11-A-1            1886.049                   18.52   
42      S355 G11-A-2            5001.665                   17.10   
57        SAE 1012-1             129.499                    8.66   
58        SAE 1018-1            3862.805                   10.04   
74         SWRY 11-1           14536.372                    2.45   

    FeSiMn Target (KG/Ton)  FeSiMn Overshot (KG/Ton)  
0                   6.1500                    1.6400  
2                   6.3550                    1.7850  
8                   1.5375                    1.6025  
10                  6.6625                    1.2275  
11                  6.6625                    1.0675  
12                  5.9450                    2.0750  
13                  8.4050                    0.5050  
20                  6.0475                    1.8725  
21                  6.0475                    1.9525  
22                  7.5850                    2.7850  
23                  6.0475                    1.9325  
24                  6.0475                    1.8625  
25                  7.5850                    0.4550  
26                  7.5850                    0.2450  
27                  9.2250                    2.6350  
28                  6.0475                    1.6725  
29                  6.0475                    1.9225  
33                  9.2250                    0.6150  
34                  6.0475                    1.6225  
35                  9.2250                    1.3450  
36                  9.2250                    1.0950  
37                 13.9400                    0.7000  
38                 13.9400                    0.7600  
39                 13.9400                    0.4800  
40                 13.9400                    0.5900  
41                 16.8100                    1.7100  
42                 16.8100                    0.2900  
57                  6.6625                    1.9975  
58                  8.4050                    1.6350  
74                  1.5375                    0.9125

In [19]:


fesi_data


SteelGrade  Billet Mass (Tons)  FeSi Actual (KG/Ton)  \
10   AISI 1012-1             127.478                  2.76   
11   AISI 1012-2            2847.713                  2.27   
16   AISI 1045-1            2321.857                  1.57   
17   AISI 1065-1            4813.276                  1.45   
18   AISI 1067-1            1677.066                  1.35   
19   AISI 1070-1             148.695                  1.22   
30       EM12K-1             125.307                  2.22   
35   S235 JR-A-1            4384.918                  3.55   
37   S275 JR-A-1           23003.444                  2.24   
41  S355 G11-A-1            1886.049                  2.07   
43   S355 J2-A-1            5487.360                  1.88   
48   S355 JR-B-1            1957.463                  1.84   
56    SAE 1008-2           20128.323                  1.50   
57    SAE 1012-1             129.499                  2.19   
63    SWRH 42B-1             279.843                  1.69   
65    SWRH 47B-1             617.338                  1.32   
66    SWRH 52B-1             294.495                  1.82   
67    SWRH 57B-1             319.415                  1.52   
69    SWRH 62B-1            2024.886                  1.29   
71    SWRH 67B-1             310.349                  1.32   

    FeSi Target (KG/Ton)  FeSi Overshot (KG/Ton)  
10                   2.0                    0.76  
11                   2.0                    0.27  
16                   1.0                    0.57  
17                   1.0                    0.45  
18                   1.0                    0.35  
19                   1.0                    0.22  
30                   2.0                    0.22  
35                   2.8                    0.75  
37                   2.1                    0.14  
41                   1.6                    0.47  
43                   1.3                    0.58  
48                   1.6                    0.24  
56                   1.3                    0.20  
57                   2.0                    0.19  
63                   1.0                    0.69  
65                   1.0                    0.32  
66                   1.0                    0.82  
67                   1.0                    0.52  
69                   1.0                    0.29  
71                   1.0                    0.32

In [20]:


fesi_data_conservative


SteelGrade  Billet Mass (Tons)  FeSi Actual (KG/Ton)  \
10   AISI 1012-1             127.478                  2.76   
11   AISI 1012-2            2847.713                  2.27   
16   AISI 1045-1            2321.857                  1.57   
17   AISI 1065-1            4813.276                  1.45   
18   AISI 1067-1            1677.066                  1.35   
19   AISI 1070-1             148.695                  1.22   
30       EM12K-1             125.307                  2.22   
35   S235 JR-A-1            4384.918                  3.55   
37   S275 JR-A-1           23003.444                  2.24   
41  S355 G11-A-1            1886.049                  2.07   
43   S355 J2-A-1            5487.360                  1.88   
48   S355 JR-B-1            1957.463                  1.84   
56    SAE 1008-2           20128.323                  1.50   
57    SAE 1012-1             129.499                  2.19   
63    SWRH 42B-1             279.843                  1.69   
65    SWRH 47B-1             617.338                  1.32   
66    SWRH 52B-1             294.495                  1.82   
67    SWRH 57B-1             319.415                  1.52   
69    SWRH 62B-1            2024.886                  1.29   
71    SWRH 67B-1             310.349                  1.32   

    FeSi Target (KG/Ton)  FeSi Overshot (KG/Ton)  
10                2.0500                  0.7100  
11                2.0500                  0.2200  
16                1.0250                  0.5450  
17                1.0250                  0.4250  
18                1.0250                  0.3250  
19                1.0250                  0.1950  
30                2.0500                  0.1700  
35                2.8700                  0.6800  
37                2.1525                  0.0875  
41                1.6400                  0.4300  
43                1.3325                  0.5475  
48                1.6400                  0.2000  
56                1.3325                  0.1675  
57                2.0500                  0.1400  
63                1.0250                  0.6650  
65                1.0250                  0.2950  
66                1.0250                  0.7950  
67                1.0250                  0.4950  
69                1.0250                  0.2650  
71                1.0250                  0.2950

In [21]:


all_in_df2 = fesi_data_conservative['SteelGrade'].isin(fesi_data['SteelGrade']).all()
all_in_df2



True

In [22]:

mean_cost_per_ton_FeSiMn = 1198.290427122754

# get consumption per kg by multiplying with kg/ton of produced billet mass:
fesmin_data_conservative["FeSiMn Overshot (KG)"] = fesmin_data_conservative["FeSiMn Overshot (KG/Ton)"]  * fesmin_data_conservative["Billet Mass (Tons)"] 

# get consumption in tons by dividing kg by 1000:
fesmin_data_conservative["FeSiMn Overshot (Tons)"] = fesmin_data_conservative["FeSiMn Overshot (KG)"]/1000

# calculate the cost of over shot by multiplying the tons used buy the cost per ton:
fesmin_data_conservative["FeSiMn Overshot (Cost$)"] = fesmin_data_conservative["FeSiMn Overshot (Tons)"] * mean_cost_per_ton_FeSiMn


fesmin_data_conservative


SteelGrade  Billet Mass (Tons)  FeSiMn Actual (KG/Ton)  \
0          3SP CON-1           10040.952                    7.79   
2          3SP UIS-1           19396.407                    8.14   
8    AISI 1006 CAQ-1           14217.135                    3.14   
10       AISI 1012-1             127.478                    7.89   
11       AISI 1012-2            2847.713                    7.73   
12       AISI 1015-1            3129.471                    8.02   
13       AISI 1018-1            8659.960                    8.91   
20    BS-G500B R22-1           11144.716                    7.92   
21    BS-G500B R22-2           11956.901                    8.00   
22    BS-G500B R51-1            1242.724                   10.37   
23      BSG500BR41-1           74008.489                    7.98   
24      BSG500BR41-2          179322.080                    7.91   
25      BSG500BR42-1            1980.089                    8.04   
26      BSG500BR42-2          151789.712                    7.83   
27      BSSS560R12-1            8578.967                   11.86   
28      BSSS560R22-1            2217.466                    7.72   
29      BSSS560R22-2            7713.394                    7.97   
33  KWSB500BWR R12-1            1364.831                    9.84   
34  KWSB500BWR R22-1           14175.175                    7.67   
35       S235 JR-A-1            4384.918                   10.57   
36       S235 JR-A-2           19308.201                   10.32   
37       S275 JR-A-1           23003.444                   14.64   
38       S275 JR-A-2           75502.051                   14.70   
39     S275 JR-AAA-1           12155.000                   14.42   
40     S275 JR-AAA-2          112660.245                   14.53   
41      S355 G11-A-1            1886.049                   18.52   
42      S355 G11-A-2            5001.665                   17.10   
57        SAE 1012-1             129.499                    8.66   
58        SAE 1018-1            3862.805                   10.04   
74         SWRY 11-1           14536.372                    2.45   

    FeSiMn Target (KG/Ton)  FeSiMn Overshot (KG/Ton)  FeSiMn Overshot (KG)  \
0                   6.1500                    1.6400          16467.161280   
2                   6.3550                    1.7850          34622.586495   
8                   1.5375                    1.6025          22782.958838   
10                  6.6625                    1.2275            156.479245   
11                  6.6625                    1.0675           3039.933628   
12                  5.9450                    2.0750           6493.652325   
13                  8.4050                    0.5050           4373.279800   
20                  6.0475                    1.8725          20868.480710   
21                  6.0475                    1.9525          23345.849203   
22                  7.5850                    2.7850           3460.986340   
23                  6.0475                    1.9325         143021.404993   
24                  6.0475                    1.8625         333987.374000   
25                  7.5850                    0.4550            900.940495   
26                  7.5850                    0.2450          37188.479440   
27                  9.2250                    2.6350          22605.578045   
28                  6.0475                    1.6725           3708.711885   
29                  6.0475                    1.9225          14828.999965   
33                  9.2250                    0.6150            839.371065   
34                  6.0475                    1.6225          22999.221438   
35                  9.2250                    1.3450           5897.714710   
36                  9.2250                    1.0950          21142.480095   
37                 13.9400                    0.7000          16102.410800   
38                 13.9400                    0.7600          57381.558760   
39                 13.9400   

In [23]:

fesmin_data_conservative["FeSiMn Overshot (Cost$)"].sum()


1094079.8834078906

In [24]:

mean_cost_per_ton_FeSi = 1718.069471192685

# get consumption per kg by multiplying with kg/ton of produced billet mass:
fesi_data_conservative["FeSi Overshot (KG)"] = fesi_data_conservative["FeSi Overshot (KG/Ton)"]  * fesi_data_conservative["Billet Mass (Tons)"]

# get consumption in tons by dividing kg by 1000:
fesi_data_conservative["FeSi Overshot (Tons)"] = fesi_data_conservative["FeSi Overshot (KG)"]/1000

# calculate the cost of over shot by multiplying the tons used buy the cost per ton:
fesi_data_conservative["FeSi Overshot (Cost$)"] = fesi_data_conservative["FeSi Overshot (Tons)"] * mean_cost_per_ton_FeSi

fesi_data_conservative


SteelGrade  Billet Mass (Tons)  FeSi Actual (KG/Ton)  \
10   AISI 1012-1             127.478                  2.76   
11   AISI 1012-2            2847.713                  2.27   
16   AISI 1045-1            2321.857                  1.57   
17   AISI 1065-1            4813.276                  1.45   
18   AISI 1067-1            1677.066                  1.35   
19   AISI 1070-1             148.695                  1.22   
30       EM12K-1             125.307                  2.22   
35   S235 JR-A-1            4384.918                  3.55   
37   S275 JR-A-1           23003.444                  2.24   
41  S355 G11-A-1            1886.049                  2.07   
43   S355 J2-A-1            5487.360                  1.88   
48   S355 JR-B-1            1957.463                  1.84   
56    SAE 1008-2           20128.323                  1.50   
57    SAE 1012-1             129.499                  2.19   
63    SWRH 42B-1             279.843                  1.69   
65    SWRH 47B-1             617.338                  1.32   
66    SWRH 52B-1             294.495                  1.82   
67    SWRH 57B-1             319.415                  1.52   
69    SWRH 62B-1            2024.886                  1.29   
71    SWRH 67B-1             310.349                  1.32   

    FeSi Target (KG/Ton)  FeSi Overshot (KG/Ton)  FeSi Overshot (KG)  \
10                2.0500                  0.7100           90.509380   
11                2.0500                  0.2200          626.496860   
16                1.0250                  0.5450         1265.412065   
17                1.0250                  0.4250         2045.642300   
18                1.0250                  0.3250          545.046450   
19                1.0250                  0.1950           28.995525   
30                2.0500                  0.1700           21.302190   
35                2.8700                  0.6800         2981.744240   
37                2.1525                  0.0875         2012.801350   
41                1.6400                  0.4300          811.001070   
43                1.3325                  0.5475         3004.329600   
48                1.6400                  0.2000          391.492600   
56                1.3325                  0.1675         3371.494102   
57                2.0500                  0.1400           18.129860   
63                1.0250                  0.6650          186.095595   
65                1.0250                  0.2950          182.114710   
66                1.0250                  0.7950          234.123525   
67                1.0250                  0.4950          158.110425   
69                1.0250                  0.2650          536.594790   
71                1.0250                  0.2950           91.552955   

    FeSi Overshot (Tons)  FeSi Overshot (Cost$)  
10              0.090509             155.501403  
11              0.626497            1076.365129  
16              1.265412            2174.065837  
17              2.045642            3514.555585  
18              0.545046             936.427666  
19              0.028996              49.816326  
30              0.021302              36.598642  
35              2.981744            5122.843750  
37              2.012801            3458.132551  
41              0.811001            1393.356179  
43              3.004330            5161.646967  
48              0.391493             672.611484  
56              3.371494            5792.461090  
57              0.018130              31.148359  
63              0.186096             319.725160  
65              0.182115             312.885724  
66              0.234124             402.240481  
67              0.158110             271.644694  
69              0.536595             921.907127  
71              0.091553             157.294337

In [25]:

fesi_data_conservative["FeSi Overshot (Cost$)"].sum()


31961.228491789516

### Overall cost:

In [27]:


fesmin_data_conservative["FeSiMn Overshot (Cost$)"].sum() + fesi_data_conservative["FeSi Overshot (Cost$)"].sum() 



1126041.11189968

In [28]:


fesmin_data_conservative["FeSiMn Overshot (Tons)"].sum()


913.0339846200006

In [29]:

fesi_data_conservative["FeSi Overshot (Tons)"].sum()


18.60298959250001

### Overall (Tons) overshot:

In [30]:

fesmin_data_conservative["FeSiMn Overshot (Tons)"].sum() + fesi_data_conservative["FeSi Overshot (Tons)"].sum()


931.6369742125006

### Merge two chemicals results, fill nan and sum results:

In [31]:


final_results = pd.merge(fesi_data_conservative, fesmin_data_conservative, on=['SteelGrade', 'Billet Mass (Tons)'],how="outer")

# to ensure that you won't get null result from summing two columns that one include null and the other is int you will get null value and it will lead to misleading results:
final_results.fillna(0.0, inplace=True)




In [32]:

final_results["Ferroalloys Overshot (KG/Ton)"] = final_results["FeSiMn Overshot (KG/Ton)"] + final_results["FeSi Overshot (KG/Ton)"]
final_results["Ferroalloys Overshot (Tons)"] =  final_results["FeSiMn Overshot (Tons)"] + final_results["FeSi Overshot (Tons)"]
final_results["Ferroalloys Overshot Cost ($USD)"] = final_results["FeSiMn Overshot (Cost$)"] + final_results["FeSi Overshot (Cost$)"]


In [33]:

fesmin_data_conservative["FeSiMn Overshot (Cost$)"].sum() + fesi_data_conservative["FeSi Overshot (Cost$)"].sum()



1126041.11189968

In [34]:

fesmin_data_conservative["FeSiMn Overshot (Tons)"].sum() + fesi_data_conservative["FeSi Overshot (Tons)"].sum()


931.6369742125006

In [35]:


final_results["FeSi Overshot (Cost$)"].sum() + final_results["FeSiMn Overshot (Cost$)"].sum()


1126041.1118996798

In [36]:

final_results["FeSi Overshot (Tons)"].sum() + final_results["FeSiMn Overshot (Tons)"].sum()


931.6369742125007

In [37]:

final_results["Ferroalloys Overshot (Tons)"].sum()



931.6369742125007

In [38]:

final_results["Ferroalloys Overshot Cost ($USD)"].sum()


1126041.11189968

### Sum numerical columns and create "total" row:

In [39]:

totals = final_results.select_dtypes(include=[np.number]).sum().rename('Total')
totals['SteelGrade'] = 'Total'
final_results = pd.concat([final_results, pd.DataFrame([totals])], ignore_index=True)
final_results


SteelGrade  Billet Mass (Tons)  FeSi Actual (KG/Ton)  \
0          3SP CON-1           10040.952                  0.00   
1          3SP UIS-1           19396.407                  0.00   
2    AISI 1006 CAQ-1           14217.135                  0.00   
3        AISI 1012-1             127.478                  2.76   
4        AISI 1012-2            2847.713                  2.27   
5        AISI 1015-1            3129.471                  0.00   
6        AISI 1018-1            8659.960                  0.00   
7        AISI 1045-1            2321.857                  1.57   
8        AISI 1065-1            4813.276                  1.45   
9        AISI 1067-1            1677.066                  1.35   
10       AISI 1070-1             148.695                  1.22   
11    BS-G500B R22-1           11144.716                  0.00   
12    BS-G500B R22-2           11956.901                  0.00   
13    BS-G500B R51-1            1242.724                  0.00   
14      BSG500BR41-1           74008.489                  0.00   
15      BSG500BR41-2          179322.080                  0.00   
16      BSG500BR42-1            1980.089                  0.00   
17      BSG500BR42-2          151789.712                  0.00   
18      BSSS560R12-1            8578.967                  0.00   
19      BSSS560R22-1            2217.466                  0.00   
20      BSSS560R22-2            7713.394                  0.00   
21           EM12K-1             125.307                  2.22   
22  KWSB500BWR R12-1            1364.831                  0.00   
23  KWSB500BWR R22-1           14175.175                  0.00   
24       S235 JR-A-1            4384.918                  3.55   
25       S235 JR-A-2           19308.201                  0.00   
26       S275 JR-A-1           23003.444                  2.24   
27       S275 JR-A-2           75502.051                  0.00   
28     S275 JR-AAA-1           12155.000                  0.00   
29     S275 JR-AAA-2          112660.245                  0.00   
30      S355 G11-A-1            1886.049                  2.07   
31      S355 G11-A-2            5001.665                  0.00   
32       S355 J2-A-1            5487.360                  1.88   
33       S355 JR-B-1            1957.463                  1.84   
34        SAE 1008-2           20128.323                  1.50   
35        SAE 1012-1             129.499                  2.19   
36        SAE 1018-1            3862.805                  0.00   
37        SWRH 42B-1             279.843                  1.69   
38        SWRH 47B-1             617.338                  1.32   
39        SWRH 52B-1             294.495                  1.82   
40        SWRH 57B-1             319.415                  1.52   
41        SWRH 62B-1            2024.886                  1.29   
42        SWRH 67B-1             310.349                  1.32   
43         SWRY 11-1           14536.372                  0.00   
44             Total          836849.582                 37.07   

    FeSi Target (KG/Ton)  FeSi Overshot (KG/Ton)  FeSi Overshot (KG)  \
0                 0.0000                  0.0000            0.000000   
1                 0.0000                  0.0000            0.000000   
2                 0.0000                  0.0000            0.000000   
3                 2.0500                  0.7100           90.509380   
4                 2.0500                  0.2200          626.496860   
5                 0.0000                  0.0000            0.000000   
6                 0.0000                  0.0000            0.000000   
7                 1.0250                  0.5450         1265.412065   
8                 1.0250                  0.4250         2045.642300   
9                 1.0250                  0.3250          545.046450   
10                1.0250                  0.1950           28.995525   
11                0.0000                  0.0000            0.000000   
12                0.0000             

In [ ]:
numerical_columns = ['Billet Mass (Tons)',
                     'FeSi Actual (KG/Ton)',
                     'FeSi Target (KG/Ton)',
                     'FeSi Overshot (KG/Ton)',
                     'FeSi Overshot (KG)',
                     'FeSi Overshot (Tons)',
                     'FeSi Overshot (Cost$)',
                     'FeSiMn Actual (KG/Ton)',
                     'FeSiMn Target (KG/Ton)',
                     'FeSiMn Overshot (KG/Ton)',
                     'FeSiMn Overshot (KG)',
                     'FeSiMn Overshot (Tons)',
                     'FeSiMn Overshot (Cost$)',
                     'Ferroalloys Overshot (KG/Ton)',
                     'Ferroalloys Overshot (Tons)',
                     'Ferroalloys Overshot Cost ($USD)']

final_results[numerical_columns] = final_results[numerical_columns].astype(float)

numerical_df = final_results.select_dtypes(include=[np.number])


for col in numerical_columns:
    final_results[col] = final_results[col].apply(lambda x: f"{x:.2f}")

final_results[numerical_columns] = final_results[numerical_columns].astype(float)



### Validate Again:

In [246]:


fesi_data["FeSi Actual (KG/Ton)"].sum()


37.07

In [247]:

fesi_data["FeSi Target (KG/Ton)"].sum()


28.7

In [248]:

fesi_data["Billet Mass (Tons)"].sum() + fesmin_data["Billet Mass (Tons)"].sum()


874910.3799999999

In [253]:

fesi_data["FeSi Overshot (KG/Ton)"].sum() + fesmin_data["FeSiMn Overshot (KG/Ton)"].sum()



55.72

In [252]:

fesmin_data["FeSiMn Overshot (Tons)"].sum() + fesi_data["FeSi Overshot (Tons)"].sum()


1103.84490678

In [254]:

fesmin_data["FeSiMn Overshot (Cost$)"].sum() + fesi_data["FeSi Overshot (Cost$)"].sum()


1333946.498867269

In [245]:


final_results[final_results["SteelGrade"] == "Total"]


SteelGrade  Billet Mass (Tons)  FeSi Actual (KG/Ton)  FeSi Target (KG/Ton)  \
46      Total           842531.28                 37.07                  28.7   

    FeSi Overshot (KG/Ton)  FeSi Overshot (KG)  FeSi Overshot (Tons)  \
46                    8.37            21585.55                 21.59   

    FeSi Overshot (Cost$)  FeSiMn Actual (KG/Ton)  FeSiMn Target (KG/Ton)  \
46               37085.47                  309.45                   262.1   

    FeSiMn Overshot (KG/Ton)  FeSiMn Overshot (KG)  FeSiMn Overshot (Tons)  \
46                     47.35            1082259.36                 1082.26   

    FeSiMn Overshot (Cost$)  Ferroalloys Overshot (KG/Ton)  \
46               1296861.03                          55.72   

    Ferroalloys Overshot (Tons)  Ferroalloys Overshot Cost ($USD)  
46                      1103.84                         1333946.5

In [256]:
fesmin_data["FeSiMn Overshot (Cost$)"].sum()




1296861.0326572799

In [257]:

fesi_data["FeSi Overshot (Cost$)"].sum()


37085.46620998909



SMP2 2023 Base estimates:
FeSiMn overshot Cost: 1296861 USD 
FeSi Overshot Cost: 37085 USD 
Overall Cost: 1333946 USD 
Ferroalloys Overshot (Tons): 1103

